// Scheduler Sync Protocol v2
//
// Bidirectional synchronization protocol between vehicle (onboard) and
// cloud (offboard) scheduler components.
//
// Key features:
// - Version vectors for conflict detection (no wall-clock dependency)
// - Source authority for deterministic conflict resolution
// - Append-only execution history
// - Offline support on both sides
//
// See docs/scheduler-sync-protocol-v2.md for full specification.

syntax = "proto3";

package swdv.scheduler_sync_v2;

option cc_enable_arenas = true;

// =============================================================================
// Version Vector
// =============================================================================

// Two-component version vector for conflict detection.
// Each component is incremented by its respective side on any change.
message JobVersion {
    uint64 cloud_seq = 1;       // Incremented by cloud on any change
    uint64 vehicle_seq = 2;     // Incremented by vehicle on any change
}

// =============================================================================
// Enumerations
// =============================================================================

// Who is authoritative for conflicts on this job.
// Set at creation time, immutable thereafter.
enum JobAuthority {
    AUTHORITY_UNKNOWN = 0;
    AUTHORITY_CLOUD = 1;        // Cloud wins conflicts (job created by cloud)
    AUTHORITY_VEHICLE = 2;      // Vehicle wins conflicts (job created on vehicle/phone)
}

// Synchronization state between cloud and vehicle.
enum SyncState {
    SYNC_STATE_UNKNOWN = 0;
    SYNC_STATE_PENDING = 1;     // Change made locally, not yet confirmed by remote
    SYNC_STATE_SYNCED = 2;      // Both sides agree on version (steady state)
    SYNC_STATE_CONFLICT = 3;    // Conflict detected, resolution in progress
}

// Job execution status.
enum JobStatus {
    JOB_STATUS_UNKNOWN = 0;
    JOB_STATUS_PENDING = 1;     // Waiting to execute
    JOB_STATUS_RUNNING = 2;     // Currently executing
    JOB_STATUS_COMPLETED = 3;   // Finished successfully
    JOB_STATUS_FAILED = 4;      // Execution failed
    JOB_STATUS_CANCELLED = 5;   // Cancelled by user/system
    JOB_STATUS_PAUSED = 6;      // Temporarily paused
}

// Vehicle wake behavior for scheduled jobs.
enum WakePolicy {
    WAKE_POLICY_NO_WAKE = 0;            // Only run if vehicle already awake
    WAKE_POLICY_WAKE_REQUIRED = 1;      // Wake vehicle via RTC to run job
}

// Vehicle sleep behavior during job execution.
enum SleepPolicy {
    SLEEP_POLICY_NORMAL = 0;            // Normal sleep after job
    SLEEP_POLICY_INHIBIT_UNTIL_COMPLETE = 1;  // Prevent sleep until done
}

// =============================================================================
// Job Record
// =============================================================================

// Complete job record with version tracking.
message JobRecord {
    // --- Identity ---
    string job_id = 1;                  // Namespaced: "cloud-*", "veh-VIN-*", "phone-*"
    JobAuthority authority = 2;         // Who wins conflicts (immutable after creation)

    // --- Version ---
    JobVersion version = 3;             // Version vector
    SyncState sync_state = 4;           // Current sync state

    // --- Lifecycle ---
    bool deleted = 5;                   // Soft delete (tombstone)
    uint64 deleted_at_ms = 6;           // When deleted (for tombstone expiry)

    // --- Content (job definition) ---
    string title = 10;                  // Human-readable title
    string service = 11;                // Target service name
    string method = 12;                 // Target method name
    string parameters_json = 13;        // JSON-encoded parameters
    uint64 scheduled_time_ms = 14;      // When to run (epoch ms)
    string recurrence_rule = 15;        // iCal RRULE for recurring jobs
    uint64 end_time_ms = 16;            // Stop recurring after this (0 = forever)

    // --- Execution state (vehicle-authoritative) ---
    JobStatus status = 20;              // Current execution status
    uint64 next_run_time_ms = 21;       // Next scheduled run
    uint64 last_executed_ms = 22;       // Last execution timestamp

    // --- Power management ---
    WakePolicy wake_policy = 25;        // Whether to wake vehicle
    SleepPolicy sleep_policy = 26;      // Sleep behavior during execution
    uint32 wake_lead_time_s = 27;       // Seconds before scheduled_time to wake

    // --- Metadata ---
    uint64 created_at_ms = 30;          // Creation timestamp
    uint64 updated_at_ms = 31;          // Last update timestamp
    string created_by = 32;             // User/system that created job
}

// =============================================================================
// Execution Record
// =============================================================================

// Immutable record of a job execution.
// Executions are append-only facts - they never conflict.
message ExecutionRecord {
    string execution_id = 1;            // Globally unique
    string job_id = 2;                  // Job that was executed
    uint64 executed_at_ms = 3;          // When execution started
    uint64 duration_ms = 4;             // Execution duration
    JobStatus status = 5;               // Result: COMPLETED or FAILED
    string result_json = 6;             // JSON result (for COMPLETED)
    string error_message = 7;           // Error message (for FAILED)
}

// =============================================================================
// Sync Messages
// =============================================================================

// Cloud → Vehicle sync message.
message C2V_SyncMessage {
    string vehicle_id = 1;              // Target vehicle
    repeated JobRecord jobs = 2;        // Job records (full state)
    repeated string deleted_job_ids = 3; // Tombstones (job_id only, for expiry)
    uint64 sync_timestamp_ms = 4;       // When this message was created
    string sync_id = 5;                 // Unique ID for this sync (for ack)
}

// Vehicle → Cloud sync message.
message V2C_SyncMessage {
    string vehicle_id = 1;              // Source vehicle
    string bridge_instance_id = 2;      // Sync bridge instance (for restart detection)
    repeated JobRecord jobs = 3;        // Job records (full state)
    repeated string deleted_job_ids = 4; // Tombstones
    repeated ExecutionRecord executions = 5; // New executions since last sync
    uint64 sync_timestamp_ms = 6;       // When this message was created
    string sync_id = 7;                 // Unique ID for this sync (for ack)
}

// Conflict resolution record (for visibility/debugging).
message ConflictResolution {
    string job_id = 1;                  // Job that had conflict
    JobVersion local_version = 2;       // Local version before resolution
    JobVersion remote_version = 3;      // Remote version before resolution
    JobVersion resolved_version = 4;    // Merged version after resolution
    string winner = 5;                  // "cloud" or "vehicle"
    string reason = 6;                  // Why this side won (e.g., "authority")
}

// Sync acknowledgment.
message SyncAck {
    string sync_id = 1;                 // ID of sync being acknowledged
    bool success = 2;                   // Overall success
    repeated ConflictResolution conflicts = 3; // Conflicts that were resolved
    uint64 ack_timestamp_ms = 4;        // When ack was created
    string error_message = 5;           // Error details if !success
}

// =============================================================================
// Sync Control Messages
// =============================================================================

// Request full sync (e.g., after long offline period or checksum mismatch).
message FullSyncRequest {
    string vehicle_id = 1;
    string reason = 2;                  // Why full sync is needed
}

// Sync status query.
message SyncStatusRequest {
    string vehicle_id = 1;
}

message SyncStatusResponse {
    string vehicle_id = 1;
    uint32 job_count = 2;               // Number of active jobs
    uint32 pending_sync_count = 3;      // Jobs waiting for sync confirmation
    uint64 last_sync_ms = 4;            // Last successful sync timestamp
    uint32 state_checksum = 5;          // CRC32 of current state (for validation)
}
